# What is this app for?

Hello!

First of all, thanks for checking out my web application, developed as part of my Master's thesis project which I conducted at Lundbeck A/S in Copenhagen, Denmark.

The web application was developed as an extension to the tools CELLEX (https://github.com/perslab/CELLEX) and CELLECT (https://github.com/perslab/CELLECT). These two tools are used together (CELLECT utilizes the output generated by CELLEX) to identify cell types which are involved in the etiology of a given disease. CELLEX takes as input a gene expression dataset (for instance a single-cell RNA-seq dataset) and calculates cell-type Expression Specificity (ES) profiles for the various cell types present in the expression data - CELLEX accomplishes this by using a "wisdom of the crowd" approach where four different measures of expression specificity are combined into a single metric, to obtain a more robust measure of expression specificity (explained more in the detail in the following paper: https://doi.org/10.7554/eLife.55851). Then, CELLECT combines these ES profiles with GWAS summary statistics (using heritability analysis models such as S-LDSC and MAGMA) to assign a significance value and thus prioritize each cell type in regards to their potential involvement in disease. The tools are explained more in depth on their respective GitHub repository sites (see the first two links above).

The web application works by taking as input the file with ES profiles (created by CELLEX) and the output folder from running the CELLECT tool (via the MAGMA pipeline, also referred to as CELLECT-MAGMA), displays various plots and tables which can help the user better understand the heritability and expression specificity of the genes from the expression data. <u>As such, the main objective of the web application is to provide the user with an easier way to explore and identify genes which might be important to disease</u>.

